### PR TITLE
Fix: unify build/run pto-isa version in CI

### LINKED ISSUE #1077: [Code Health] CI: onboard a2a3 host_runtime built against unpinned pto-isa diverges from test-time --pto-isa-commit
### Category

Robustness (potential edge-case failure)

### Component

Build System

### Description

CI 在同一次 onboard a2a3 run 里会用**两个不同的 pto-isa 版本**：

- **Build 期**（`pip install` → CMake `build_runtimes` target → `build_runtimes.py`）：对 a2a3 调用 `ensure_pto_isa_root(clone_protocol=...)` **不传 commit**，所以 clone 的是 pto-isa **HEAD**（或 self-hosted runner 上一轮残留的 clone）。`host_runtime.so` 在这一步把 `comm_hccl.cpp` 里 `#include "pto/npu/comm/async/sdma/sdma_workspace_manager.hpp"` 编进去，scene test 直接加载 `build/lib/` 的预编译产物、不会重编。
- **Run 期**（scene test）：用 `--pto-isa-commit ${{ env.PTO_ISA_COMMIT }}` pin 到固定 commit，conftest 把 `build/pto-isa` checkout 到该 commit，**只影响 kernel 编译**。

结果：`host_runtime.so`（内嵌 pto-isa SDMA 头）与测试期 kernel 用了两个不同的 pto-isa 修订版。调用失败符号的 `SdmaWorkspaceManager` 整体住在 `host_runtime.so` 里，真正跑的是 build 期那版 pto-isa，而非 pin 的版本。

**迫近风险**：pto-isa HEAD 已经把 `sdma_workspace_manager.hpp` 从 `pto/npu/comm/async/sdma/` 移到 `pto/comm/async/sdma/`（pto-isa commit `8e0400f2`）。一旦 build 期 fresh clone 到不含旧路径的 HEAD，a2a3 host 会因 `#include "pto/npu/comm/..."` 找不到头文件而**直接编译失败**。

这是在排查 #1067（`aclnnShmemSdmaStarsQueryGetWorkspaceSize failed`）时发现的**独立** build-hygiene 问题。已确认它**不是** #1067 的根因（#1067 是设备侧 CANN op 运行期失败；build 期 pin 版到 HEAD 之间对该 header 的改动是行为等价的 refactor），但属于应当单独修复的 CI 一致性缺陷。

附带：`src/a2a3/platform/onboard/host/comm_hccl.cpp` 的 `ensure_sdma_workspace` 注释声称 workspace init 失败时 demo 会 "self-skip"，但对应 demo test 并无 self-skip 逻辑——独立小问题，可在 #1067 一并处理，此处仅记录。

Git Commit ID: c4b0aac2c8b17d2748793337d47df2f2a1552127

Related: #1067

### Location

- `simpler_setup/build_runtimes.py` — a2a3 块 `ensure_pto_isa_root(...)` 调用未传 `commit`
- `CMakeLists.txt` — `build_runtimes` custom target 未向脚本透传 pinned commit
- `.github/workflows/ci.yml` — onboard job 的 `pip install` 未 pin，仅 scene test 步骤用 `--pto-isa-commit ${{ env.PTO_ISA_COMMIT }}`
- `src/a2a3/platform/onboard/host/comm_hccl.cpp` — `#include "pto/npu/comm/async/sdma/sdma_workspace_manager.hpp"`（依赖已被 pto-isa HEAD 移走的路径）

### Proposed Fix

让 build 期与 run 期共用唯一的 pinned commit（ci.yml 的 `env.PTO_ISA_COMMIT` 作为单一源头，改一处两端同步）：

1. `build_runtimes.py` 新增 `--pto-isa-commit`，透传给 `ensure_pto_isa_root(commit=...)`。
2. `CMakeLists.txt` 新增 `SIMPLER_PTO_ISA_COMMIT` cache 变量（仿 `SIMPLER_PTO_CLONE_PROTOCOL`），非空时把 `--pto-isa-commit <sha>` 传给 build_runtimes target。
3. `ci.yml` onboard job 的 `pip install` 加 `--config-settings=cmake.define.SIMPLER_PTO_ISA_COMMIT="${PTO_ISA_COMMIT}"`；`requires_hardware` 的 pytest 也加 `--pto-isa-commit`，避免 conftest 在无 pin 时把 `build/pto-isa` reset 回 HEAD。

### Priority

High (significant risk, fix soon)

### PR DESCRIPTION
**Pull request by ChaoZheng109**
## Summary

Onboard a2a3 `host_runtime.so` embeds the pto-isa SDMA headers
(`comm_hccl.cpp`'s `#include "pto/npu/comm/.../sdma_workspace_manager.hpp"`),
but it was built during `pip install` against an **unpinned** pto-isa (HEAD,
or a stale clone left on the self-hosted runner), while scene tests pin
pto-isa via `--pto-isa-commit`. So in one CI run `host_runtime.so` and the
test-time kernels could be built against **two different pto-isa revisions**.
Because pto-isa HEAD has since moved `sdma_workspace_manager.hpp` out of the
`pto/npu/comm/...` path, an unpinned fresh clone would now also fail to
compile the a2a3 host outright.

This makes both ends use the **single** `ci.yml` `env.PTO_ISA_COMMIT` — bump
that one line and build + run update together.

- **`build_runtimes.py`**: add `--pto-isa-commit`, pass it to
  `ensure_pto_isa_root(commit=...)` for the a2a3 host build.
- **`CMakeLists.txt`**: add `SIMPLER_PTO_ISA_COMMIT` cache var (mirrors
  `SIMPLER_PTO_CLONE_PROTOCOL`), threaded into the `build_runtimes` target
  only when set.
- **`.github/workflows/ci.yml`**: onboard `pip install` passes the commit via
  `--config-settings=cmake.define.SIMPLER_PTO_ISA_COMMIT="${PTO_ISA_COMMIT}"`;
  the `requires_hardware` pytest steps also pin `--pto-isa-commit` so
  `conftest` doesn't reset `build/pto-isa` back to HEAD before the a2a3 C++ UT
  host rebuild.
- **`docs/developer-guide.md`**: document the pto-isa-commit rebuild trigger.

Note: the pinned commit value is **unchanged** (`ddafa8da…`); this only makes
the build honor the same value the tests already use.

## Scope / non-goals

This is the build-hygiene problem found while investigating #1067 — it is
**not** that issue's root cause (which is a device-side CANN op runtime
failure of `aclnnShmemSdmaStarsQueryGetWorkspaceSize`, independent of the
pto-isa header revision). The `@pytest.mark.skip` on
`test_sdma_async_completion_demo` therefore stays in place.

## Testing

- [x] `pre-commit` hooks pass (markdownlint, ruff, pyright, yaml, headers)
- [x] CMake conditional-arg expansion verified (empty cache var → flag
      omitted, 0 stray args; set → `--pto-isa-commit <sha>`)
- [x] `python -m py_compile build_runtimes.py`, YAML parse
- [ ] Full effect is only observable on the self-hosted onboard runners
      (a2a3 / a5) — verify the onboard CI jobs build host_runtime against the
      pinned commit.

Fixes #1077
Related: #1067

### CODE REVIEW

**Comment by ChaoZheng109** [comment]
Addressed the bot review feedback (pushed as a squashed update):

- **CodeRabbit (Major) — DFX smokes bypass the pin**: ✅ Fixed. The a2a3 DFX smoke steps (`dep_gen`, `l2_swimlane`, `pmu`, `tensor_dump`) ran pytest without `--pto-isa-commit`, so `conftest`'s `ensure_pto_isa_root(update_if_exists=True)` could reset `build/pto-isa` to HEAD and reintroduce host-runtime/kernel drift within the same job. Pinned all of them — both the onboard (`ssh`) and the sim mirror (`https`) — to `env.PTO_ISA_COMMIT`.
- **Gemini (medium) — CMake `if(var)` truthiness**: ✅ Fixed. Changed to `if(NOT SIMPLER_PTO_ISA_COMMIT STREQUAL "")` and quoted the expansion, so a numeric/boolean-like SHA can't be misread.
- **Gemini (medium) — strip/validate whitespace in `build_runtimes.py`**: ⏭️ Skipped. The premise (silent `.strip()` normalization) doesn't apply — the code doesn't strip anything. The value comes from the single-source `env.PTO_ISA_COMMIT` and a malformed commit already fails fast at `git checkout`; adding speculative input validation here is gold-plating.

Note: the earlier `ut` / `ut-a2a3` red was the pre-existing `test_scope_stats_collector` duplicate-target collision on main (`c4b0aac2`), unrelated to this PR — being fixed in #1080.

File: CMakeLists.txt [hunk 1]
```diff
@@ -36,13 +36,29 @@ add_subdirectory(python/bindings)
...
 # -fsanitize token list. Enable at `pip install` time via
 #   pip install --no-build-isolation --config-settings=cmake.define.SIMPLER_SANITIZER=asan .
 set(SIMPLER_SANITIZER "none" CACHE STRING
     "Sanitizer for host targets during install (none/asan/ubsan/tsan)")
 
+# Pass --pto-isa-commit only when pinned; an empty cache var leaves the flag
+# off (build_runtimes then clones/uses HEAD).
+if(SIMPLER_PTO_ISA_COMMIT)
+    set(_pto_isa_commit_arg --pto-isa-commit ${SIMPLER_PTO_ISA_COMMIT})
+endif()
```
[the remark this answers is not in the document]
Comment (ChaoZheng109) [reply CMakeLists.txt]: ✅ Done — changed to `if(NOT SIMPLER_PTO_ISA_COMMIT STREQUAL "")` and quoted the expansion so a numeric/boolean-like SHA can't be misread by CMake's if() truthiness.

File: simpler_setup/build_runtimes.py [hunk 2]
```diff
@@ -121,10 +126,17 @@ def build_all(
...
+    # check out via `--pto-isa-commit`. Without it, host_runtime.so would be
+    # built against whatever pto-isa HEAD (or a stale clone) happened to be at
+    # install time, diverging from the test-time kernels — see issue #1067.
     if "a2a3" in platforms:
         from simpler_setup.pto_isa import ensure_pto_isa_root  # noqa: PLC0415
 
-        os.environ["PTO_ISA_ROOT"] = ensure_pto_isa_root(clone_protocol=clone_protocol, verbose=True)
+        os.environ["PTO_ISA_ROOT"] = ensure_pto_isa_root(
+            commit=pto_isa_commit, clone_protocol=clone_protocol, verbose=True
+        )
```
[the remark this answers is not in the document]
Comment (ChaoZheng109) [reply simpler_setup/build_runtimes.py:139]: Skipped — the premise doesn't apply: the code doesn't strip/normalize the commit at all (it forwards `pto_isa_commit` verbatim to `ensure_pto_isa_root`). The value comes from the single-source `env.PTO_ISA_COMMIT`, and a malformed commit already fails fast at the `git checkout` inside `ensure_pto_isa_root`. Adding speculative whitespace validation here would be gold-plating for a controlled input, so leaving it out.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -9,6 +9,11 @@ concurrency:
   cancel-in-progress: true
 
 env:
+  # Single source of truth for the pinned pto-isa revision. Drives BOTH the
+  # build (`pip install` threads it into build_runtimes.py via
+  # SIMPLER_PTO_ISA_COMMIT, so host_runtime.so embeds these pto-isa headers)
+  # and the run (scene tests check it out via `--pto-isa-commit`). Bump here to
+  # update both at once — never pin one without the other (issue #1067).
   PTO_ISA_COMMIT: ddafa8da9c760ecd13fe9fe2833d6ee55fb20bd8
```
[the remark this answers is not in the document]
Comment (ChaoZheng109) [reply .github/workflows/ci.yml:17]: ✅ Good catch — fixed. The a2a3 DFX smoke steps (dep_gen, l2_swimlane, pmu, tensor_dump) now pass `--pto-isa-commit ${{ env.PTO_ISA_COMMIT }}`, for both the onboard (ssh) and the sim mirror (https) sets, so conftest no longer resets build/pto-isa to HEAD within the same job.

**Review by ChaoWao — APPROVED** [review]
